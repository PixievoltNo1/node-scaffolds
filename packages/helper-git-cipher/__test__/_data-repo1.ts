import type { IHashAlgorithm } from '@guanghechen/helper-mac'

export * from '@guanghechen/helper-git/__test__/_data-repo1'

export const repo1CryptCommitIdTable = {
  A: 'fad3c6db42571ed6c2f48ba9e7b7f75ee3b8e328',
  B: '8edffe3781a61e1e542f4f9af8a13f58e5197c36',
  C: '90e3f12fb634343d32ec04c2fcf084083d36a54e',
  D: '99d394a1fa63f7ec8d3e9c0c59c9ba0bd1262dce',
  E: '8e8afc6b41c14321b4ccfbf755534354ec9cce05',
  F: 'a55645b3caec41a6c5384b064359c52f4b3726bd',
  G: 'd9eae9d91e0f9345e47cbc77590640edbac10ee1',
  H: 'b9ba89528832cbe2303ca9537d254558abbcfe2b',
  I: '2e247f16f893b112cc28e51eb9a8b3c49a40e4b1',
  J: 'e7b530a1796f1a206a02341fdfb5df0a772b3f35',
  K: '7df3706a51d8e62c72c83bfc22ee70b015a69e8c',
}

export const repo1CryptCommitMessageTable = {
  A: '97bd8d84dc83c508077b247206ccb696eb066828c88ba62f96a0388061189093',
  B: '8ef2579bebef6f8de168d94dc51557b696bdeadd91cd54dae8c5adaa056bd5ab',
  C: 'f56f745fe0ca5729d11b9f6245d3dd6b92ab65d0071d2f44fb87d829db0f0bf1',
  D: 'c9f30d2c4ae5f85262d7c0f937f908269ff170fa860a084c801e4043c4809777',
  E: 'e35842380c2943afbf6c488340217debe8a286f3357ab0deaafe8bae511b896f',
  F: '4686d1873916f43c116d3ce3dfd11eb271371363b1ef93274c0034f0b8479b71',
  G: '2d683f93b46b83572ecc281132deaf6e2da3ecc43935605178381631a2f1530b',
  H: '77dfd635e6db3d6d1c71df9f5cb33bce084584ffccba91d73a47a0e7f25683f9',
  I: 'c667c67de669e93c33d597deb746c9257768208a1f58bd3f1d995eccfdaa15ba',
  J: '5191065827930de1fb814932e2a6064bba4f12d0472cf1ece8e823da817615c1',
  K: '79861cb279ecfa722603ccd0d601d07d90cc1899abf8e51b0e83907b0decc0d4',
}

export const itemTable = {
  A: {
    plainFilepath: 'a.txt',
    cryptFilepath: 'a.txt',
    cryptFilepathParts: [],
    fingerprint: '4e26698e6bebd87fc210bec49fea4da6210b5769dbff50b3479effa16799120f',
    keepPlain: true,
    iv: Buffer.from('c65b6cc1c81a805ba6a7187e', 'hex'),
    authTag: undefined,
  },
  A2: {
    plainFilepath: 'a.txt',
    cryptFilepath: 'a.txt',
    cryptFilepathParts: [],
    fingerprint: '4ec33c94039179da5febb8936428e80e7b0d3f42689a4adb38fc8e479634eeb8',
    keepPlain: true,
    iv: Buffer.from('2b6d63cbde1c274dfe8ba9bc', 'hex'),
    authTag: undefined,
  },
  B: {
    plainFilepath: 'b.txt',
    cryptFilepath: 'asuna/7162c5dc69f5c2855c2fa7e454c8e4ed7e03a47c91b55be24e8931b8b099ed93',
    cryptFilepathParts: [],
    fingerprint: '965bfea36969b0b94ab0665baedd328c37f200340b937c07fdb6885ea363993c',
    keepPlain: false,
    iv: Buffer.from('44ad63f5398f8d806658f35d', 'hex'),
    authTag: Buffer.from('4ecc052246ff595a5549510b4a639105', 'hex'),
  },
  B2: {
    plainFilepath: 'b.txt',
    cryptFilepath: 'asuna/7162c5dc69f5c2855c2fa7e454c8e4ed7e03a47c91b55be24e8931b8b099ed93',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2'],
    fingerprint: '88d03f260a5158dd23220ae24160320df2ec63840dac45ad4c99cc6d0208e248',
    keepPlain: false,
    size: 1650,
    iv: Buffer.from('c2e17ce7a041aa6a34b6508a', 'hex'),
    authTag: Buffer.from('ddcc8070520f86d9ba48a43d17b4bfb5', 'hex'),
  },
  C: {
    plainFilepath: 'x/c.txt',
    cryptFilepath: 'asuna/0114c437562d21ed83066b8af1836df8783f5eaea3646ab14da6f7f4ef957083',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2', '.ghc-part3'],
    fingerprint: 'fd7dc434ab208f15cd61dcc39e8e67de75a1cc6e1c6c9268d653a01b819da054',
    keepPlain: false,
    iv: Buffer.from('ea36cfca05bee4b045956a1f', 'hex'),
    authTag: Buffer.from('dff151f0961471d5782eed60e57ed5af', 'hex'),
  },
  C2: {
    plainFilepath: 'x/c.txt',
    cryptFilepath: 'asuna/0114c437562d21ed83066b8af1836df8783f5eaea3646ab14da6f7f4ef957083',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2', '.ghc-part3', '.ghc-part4'],
    fingerprint: 'df91313a8fc51bce13227ad3b5e8eeea815fe149969c174b0f2da373dea473c1',
    keepPlain: false,
    iv: Buffer.from('38eff44fe4585a632a75464a', 'hex'),
    authTag: Buffer.from('90ec961ea6e2936020f6be60ec0b6239', 'hex'),
  },
  C3: {
    plainFilepath: 'x/c.txt',
    cryptFilepath: 'asuna/0114c437562d21ed83066b8af1836df8783f5eaea3646ab14da6f7f4ef957083',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2', '.ghc-part3', '.ghc-part4', '.ghc-part5'],
    fingerprint: 'ce9d19d55a13973f3e1f970e8915df800a1a61bd2f38212b55db73b8669ddcaf',
    keepPlain: false,
    iv: Buffer.from('362b53ef8d8ac14d4689bca3', 'hex'),
    authTag: Buffer.from('2674b106e4f472435f2b9e8ce2a6d83a', 'hex'),
  },
  D: {
    plainFilepath: 'x/d.txt',
    cryptFilepath: 'asuna/a468e223dd684ed8393c6eeb2d7e29929b890aa186f02d570a93a514ad72ebde',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2', '.ghc-part3', '.ghc-part4'],
    fingerprint: '40cb73b4c02d34812f38a5ca3a3f95d377285e83d7bb499573b918e1862bcf13',
    keepPlain: false,
    iv: Buffer.from('6a60b80f8e5c5d6772cda449', 'hex'),
    authTag: Buffer.from('333e0019fff2b8f293cdb1f698ef114e', 'hex'),
  },
  E: {
    plainFilepath: 'y/z/e.txt',
    cryptFilepath: 'asuna/ad931c88ceba4f06455236a3a58c3d8c04b7690e687ad5d4dded1b53d4d82763',
    cryptFilepathParts: ['.ghc-part1', '.ghc-part2', '.ghc-part3', '.ghc-part4'],
    fingerprint: '40cb73b4c02d34812f38a5ca3a3f95d377285e83d7bb499573b918e1862bcf13',
    keepPlain: false,
    iv: Buffer.from('6a60b80f8e5c5d6772cda449', 'hex'),
    authTag: Buffer.from('333e0019fff2b8f293cdb1f698ef114e', 'hex'),
  },
}

export const diffItemsTable = {
  stepA: [
    { changeType: 'added', newItem: itemTable.A },
    { changeType: 'added', newItem: itemTable.B },
  ],
  stepB: [
    { changeType: 'added', newItem: itemTable.C },
    { changeType: 'modified', oldItem: itemTable.A, newItem: itemTable.A2 },
  ],
  stepE: [
    { changeType: 'removed', oldItem: itemTable.B },
    { changeType: 'modified', oldItem: itemTable.C, newItem: itemTable.C2 },
  ],
  stepI: [
    { changeType: 'added', newItem: itemTable.D },
    { changeType: 'modified', oldItem: itemTable.B2, newItem: itemTable.B },
  ],
  stepK: [
    { changeType: 'removed', oldItem: itemTable.D },
    { changeType: 'added', newItem: itemTable.E },
  ],
}

export const cryptFilesDir = 'asuna'
export const maxTargetFileSize = 1024
export const partCodePrefix = '.ghc-part'
export const contentHashAlgorithm: IHashAlgorithm = 'sha256'
export const pathHashAlgorithm: IHashAlgorithm = 'sha256'
