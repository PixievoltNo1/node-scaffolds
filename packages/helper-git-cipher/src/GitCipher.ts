import type {
  FileCipherPathResolver,
  IFileCipherBatcher,
  IFileCipherCatalog,
  IJsonConfigKeeper,
} from '@guanghechen/helper-cipher-file'
import type { ILogger } from '@guanghechen/utility-types'
import { decryptGitRepo } from './decrypt/repo'
import { encryptGitRepo } from './encrypt/repo'
import type { IGitCipherConfigData } from './types'

export interface IGitCipherProps {
  cipherBatcher: IFileCipherBatcher
  configKeeper: IJsonConfigKeeper<IGitCipherConfigData>
  logger?: ILogger
  multilineMessagePrefix?: string
}

export interface IGitCipherEncryptParams {
  catalog: IFileCipherCatalog
  pathResolver: FileCipherPathResolver
}

export interface IGitCipherDecryptParams {
  pathResolver: FileCipherPathResolver
}

export class GitCipher {
  public readonly cipherBatcher: IFileCipherBatcher
  public readonly configKeeper: IJsonConfigKeeper<IGitCipherConfigData>
  public readonly logger?: ILogger
  public readonly multilineMessagePrefix?: string

  constructor(props: IGitCipherProps) {
    this.cipherBatcher = props.cipherBatcher
    this.configKeeper = props.configKeeper
    this.logger = props.logger
    this.multilineMessagePrefix = props.multilineMessagePrefix
  }

  public async encrypt(params: IGitCipherEncryptParams): Promise<void> {
    const { cipherBatcher, configKeeper, logger, multilineMessagePrefix } = this
    await encryptGitRepo({
      pathResolver: params.pathResolver,
      catalog: params.catalog,
      cipherBatcher,
      configKeeper,
      logger,
      multilineMessagePrefix,
    })
  }

  public async decrypt(params: IGitCipherDecryptParams): Promise<void> {
    const { cipherBatcher, configKeeper, logger, multilineMessagePrefix } = this
    await decryptGitRepo({
      pathResolver: params.pathResolver,
      cipherBatcher,
      configKeeper,
      logger,
      multilineMessagePrefix,
    })
  }
}
