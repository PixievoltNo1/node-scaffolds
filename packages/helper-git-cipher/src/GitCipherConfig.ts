import type { ICipher } from '@guanghechen/helper-cipher'
import type { IJsonConfigKeeper } from '@guanghechen/helper-cipher-file'
import { rm, writeFile } from '@guanghechen/helper-fs'
import invariant from '@guanghechen/invariant'
import { existsSync, statSync } from 'node:fs'
import fs from 'node:fs/promises'
import type { IGitCipherConfigData } from './types'

export interface IGitCipherConfigProps {
  readonly cipher: ICipher
  readonly filepath: string
}

export class GitCipherConfig implements IJsonConfigKeeper<IGitCipherConfigData> {
  public readonly cipher: ICipher
  public readonly filepath: string

  constructor(props: IGitCipherConfigProps) {
    this.cipher = props.cipher
    this.filepath = props.filepath
  }

  public async load(): Promise<IGitCipherConfigData | null> {
    const { cipher, filepath } = this
    if (!existsSync(filepath)) return null

    invariant(statSync(filepath).isFile(), `[GitCipherConfig.load] not a file. (${filepath})`)

    const cryptData: Buffer = await fs.readFile(filepath)
    const plainData: Buffer = cipher.decrypt(cryptData)
    const jsonContent: string = plainData.toString('utf8')
    const data = JSON.parse(jsonContent)
    return data as IGitCipherConfigData
  }

  public async save(data: IGitCipherConfigData): Promise<void> {
    const { cipher, filepath } = this
    const jsonContent: string = JSON.stringify(data)
    const plainData: Buffer = Buffer.from(jsonContent, 'utf8')
    const cryptData: Buffer = cipher.encrypt(plainData)
    await writeFile(filepath, cryptData)
  }

  public async remove(): Promise<void> {
    await rm(this.filepath)
  }
}
