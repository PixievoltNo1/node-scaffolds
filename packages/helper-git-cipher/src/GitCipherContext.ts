import type {
  ICatalogItem,
  ICipherCatalog,
  ICipherCatalogContext,
  IDeserializedCatalogItem,
} from '@guanghechen/cipher-workspace.types'
import type { IConfigKeeper } from '@guanghechen/config'
import type { IFileCipherBatcher } from '@guanghechen/helper-cipher-file'
import { FileCipherCatalog } from '@guanghechen/helper-cipher-file'
import type { IReporter } from '@guanghechen/reporter.types'
import type { IGitCipherConfig } from './types'

export interface IGitCipherContextProps {
  readonly catalogContext: ICipherCatalogContext
  readonly cipherBatcher: IFileCipherBatcher
  readonly configKeeper: IConfigKeeper<IGitCipherConfig>
  readonly reporter: IReporter
}

export interface IGitCipherContext {
  readonly catalog: ICipherCatalog
  readonly configKeeper: IConfigKeeper<IGitCipherConfig>
  readonly cipherBatcher: IFileCipherBatcher
  readonly reporter: IReporter
  flatItem(item: IDeserializedCatalogItem): Promise<ICatalogItem>
}

export class GitCipherContext implements IGitCipherContext {
  public readonly catalog: ICipherCatalog
  public readonly cipherBatcher: IFileCipherBatcher
  public readonly configKeeper: IConfigKeeper<IGitCipherConfig>
  public readonly reporter: IReporter

  constructor(props: IGitCipherContextProps) {
    const { catalogContext, cipherBatcher, configKeeper, reporter } = props
    this.catalog = new FileCipherCatalog(catalogContext)
    this.cipherBatcher = cipherBatcher
    this.configKeeper = configKeeper
    this.reporter = reporter
  }

  public readonly flatItem = async (item: IDeserializedCatalogItem): Promise<ICatalogItem> => {
    const { catalog } = this
    return {
      ...item,
      cryptFilepath: catalog.calcCryptFilepath(item.plainFilepath),
      iv: await catalog.getIv(item),
      authTag: item.authTag,
    }
  }
}
