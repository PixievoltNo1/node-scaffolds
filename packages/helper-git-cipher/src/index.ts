export * from './GitCipher'
export * from './GitCipherConfig'
export * from './types'
export * from './util'
