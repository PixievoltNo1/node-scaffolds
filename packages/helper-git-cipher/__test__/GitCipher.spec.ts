import { ChalkLogger, Level } from '@guanghechen/chalk-logger'
import { AesCipherFactory } from '@guanghechen/helper-cipher'
import {
  FileCipher,
  FileCipherBatcher,
  FileCipherCatalog,
  FileCipherPathResolver,
} from '@guanghechen/helper-cipher-file'
import { BigFileHelper } from '@guanghechen/helper-file'
import { emptyDir, rm } from '@guanghechen/helper-fs'
import type { IGitCommandBaseParams } from '@guanghechen/helper-git'
import {
  checkBranch,
  createBranch,
  deleteBranch,
  getAllLocalBranches,
  getCommitInTopology,
  getCommitWithMessageList,
} from '@guanghechen/helper-git'
import { buildRepo1 } from '@guanghechen/helper-git/__test__/_data-repo1'
import type { IConsoleMock } from '@guanghechen/helper-jest'
import { createConsoleMock } from '@guanghechen/helper-jest'
import { locateFixtures } from 'jest.helper'
import path from 'node:path'
import { GitCipher, GitCipherConfig } from '../src'

const encoding: BufferEncoding = 'utf8'
const cryptDir = 'encrypted'
const maxTargetFileSize = 1024
const partCodePrefix = '.ghc-part'
const multilineMessagePrefix = '    '

const cryptCommitIdTable = {
  A: '67b26f8e00df18a608e465a5680efd30ced8d02d',
  B: 'b5895a740051257b04f92b1b750b3ba6c709c716',
  C: '07fc2d2b3d1037cc743812c89039f26f99df925c',
  D: '4af9bd6a4700ab759d57d458fe02d335f1c86f29',
  E: '21944054759529758ffe4d6a5106509809807cb1',
  F: '8badc060f441834fa2258a5732647301ceb6d583',
  G: '0d0cab88c672480c3095a04ec24598684bb573b1',
  H: '7ceac78048361b622b67ca249c568d5b3f6386bb',
  I: '7ce10051a0885c345b511d1b479597b620fcdc95',
  J: '7324e4d828e78f4b04386a37c14fd810a57815b5',
  K: '5b43023ea85cbb6d8605a2bb1f0ac27248c9c1ec',
}

describe('GitCipher', () => {
  const workspaceDir: string = locateFixtures('__fictitious__GitCipher')
  const plainRootDir: string = path.join(workspaceDir, 'plain')
  const cryptRootDir: string = path.join(workspaceDir, 'crypt')
  const bakRootDir: string = path.join(workspaceDir, 'plain_bak')
  const pathResolver = new FileCipherPathResolver({ plainRootDir, cryptRootDir })
  const bakPathResolver = new FileCipherPathResolver({ plainRootDir: bakRootDir, cryptRootDir })
  const logger = new ChalkLogger({
    name: 'GitCipher',
    level: Level.ERROR,
    flags: { inline: true },
  })

  const plainCtx: IGitCommandBaseParams = { cwd: plainRootDir, logger, execaOptions: {} }
  const cryptCtx: IGitCommandBaseParams = { cwd: cryptRootDir, logger, execaOptions: {} }
  const bakPlainCtx: IGitCommandBaseParams = { cwd: bakRootDir, logger, execaOptions: {} }

  const fileHelper = new BigFileHelper({ partCodePrefix: partCodePrefix })
  const cipherFactory = new AesCipherFactory()
  const cipher = cipherFactory.initFromPassword(Buffer.from('guanghechen', encoding), {
    salt: 'salt',
    iterations: 100000,
    keylen: 32,
    digest: 'sha256',
  })
  const fileCipher = new FileCipher({ cipher })
  const cipherBatcher = new FileCipherBatcher({
    fileCipher,
    fileHelper,
    maxTargetFileSize,
    logger,
  })

  const configKeeper = new GitCipherConfig({
    cipher,
    filepath: path.join(cryptRootDir, 'catalog.ghc.txt'),
  })
  const gitCipher = new GitCipher({
    cipherBatcher,
    configKeeper,
    logger,
    multilineMessagePrefix,
  })

  const catalog = new FileCipherCatalog({
    pathResolver,
    cryptDir,
    maxTargetFileSize,
    partCodePrefix,
    logger,
    isKeepPlain: sourceFilepath => sourceFilepath === 'a.txt',
  })

  let logMock: IConsoleMock
  beforeEach(async () => {
    logMock = createConsoleMock(['error'])
    await emptyDir(workspaceDir)
  })
  afterEach(async () => {
    await configKeeper.remove()
    await rm(workspaceDir)
    logMock.restore()
  })

  test(
    'from scratch',
    async () => {
      const { commitIdTable, commitTable } = await buildRepo1({
        repoDir: plainRootDir,
        logger,
        execaOptions: {},
      })

      // Test encrypt.
      await gitCipher.encrypt({ catalog, pathResolver })
      expect(await getAllLocalBranches({ ...cryptCtx })).toEqual({
        currentBranch: 'main',
        branches: ['main'],
      })
      expect(await getCommitInTopology({ ...cryptCtx, branchOrCommitId: 'main' })).toEqual([
        { id: cryptCommitIdTable.A, parents: [] },
        { id: cryptCommitIdTable.B, parents: [cryptCommitIdTable.A] },
        { id: cryptCommitIdTable.C, parents: [cryptCommitIdTable.B] },
        { id: cryptCommitIdTable.D, parents: [cryptCommitIdTable.C] },
        { id: cryptCommitIdTable.E, parents: [cryptCommitIdTable.C] },
        { id: cryptCommitIdTable.F, parents: [cryptCommitIdTable.E, cryptCommitIdTable.D] },
        { id: cryptCommitIdTable.G, parents: [cryptCommitIdTable.B] },
        { id: cryptCommitIdTable.H, parents: [cryptCommitIdTable.G, cryptCommitIdTable.E] },
        { id: cryptCommitIdTable.I, parents: [cryptCommitIdTable.G] },
        {
          id: cryptCommitIdTable.J,
          parents: [cryptCommitIdTable.I, cryptCommitIdTable.H, cryptCommitIdTable.F],
        },
        { id: cryptCommitIdTable.K, parents: [cryptCommitIdTable.J] },
      ])
      expect(await getCommitWithMessageList({ ...cryptCtx, branchOrCommitIds: ['main'] })).toEqual([
        { id: cryptCommitIdTable.K, message: `#${commitIdTable.K}` },
        { id: cryptCommitIdTable.J, message: `#${commitIdTable.J}` },
        { id: cryptCommitIdTable.I, message: `#${commitIdTable.I}` },
        { id: cryptCommitIdTable.H, message: `#${commitIdTable.H}` },
        { id: cryptCommitIdTable.G, message: `#${commitIdTable.G}` },
        { id: cryptCommitIdTable.F, message: `#${commitIdTable.F}` },
        { id: cryptCommitIdTable.E, message: `#${commitIdTable.E}` },
        { id: cryptCommitIdTable.D, message: `#${commitIdTable.D}` },
        { id: cryptCommitIdTable.C, message: `#${commitIdTable.C}` },
        { id: cryptCommitIdTable.B, message: `#${commitIdTable.B}` },
        { id: cryptCommitIdTable.A, message: `#${commitIdTable.A}` },
      ])

      // Test decrypt.
      await gitCipher.decrypt({ pathResolver: bakPathResolver })
      expect(await getAllLocalBranches({ ...bakPlainCtx })).toEqual({
        currentBranch: 'main',
        branches: ['main'],
      })
      expect(await getCommitInTopology({ ...bakPlainCtx, branchOrCommitId: 'main' })).toEqual([
        { id: commitIdTable.A, parents: [] },
        { id: commitIdTable.B, parents: [commitIdTable.A] },
        { id: commitIdTable.C, parents: [commitIdTable.B] },
        { id: commitIdTable.D, parents: [commitIdTable.C] },
        { id: commitIdTable.E, parents: [commitIdTable.C] },
        { id: commitIdTable.F, parents: [commitIdTable.E, commitIdTable.D] },
        { id: commitIdTable.G, parents: [commitIdTable.B] },
        { id: commitIdTable.H, parents: [commitIdTable.G, commitIdTable.E] },
        { id: commitIdTable.I, parents: [commitIdTable.G] },
        {
          id: commitIdTable.J,
          parents: [commitIdTable.I, commitIdTable.H, commitIdTable.F],
        },
        { id: commitIdTable.K, parents: [commitIdTable.J] },
      ])
      expect(
        await getCommitWithMessageList({ ...bakPlainCtx, branchOrCommitIds: ['main'] }),
      ).toEqual([
        { id: commitIdTable.K, message: commitTable.K.message },
        { id: commitIdTable.J, message: commitTable.J.message },
        { id: commitIdTable.I, message: commitTable.I.message },
        { id: commitIdTable.H, message: commitTable.H.message },
        { id: commitIdTable.G, message: commitTable.G.message },
        { id: commitIdTable.F, message: commitTable.F.message },
        { id: commitIdTable.E, message: commitTable.E.message },
        { id: commitIdTable.D, message: commitTable.D.message },
        { id: commitIdTable.C, message: commitTable.C.message },
        { id: commitIdTable.B, message: commitTable.B.message },
        { id: commitIdTable.A, message: commitTable.A.message },
      ])

      expect(logMock.getIndiscriminateAll()).toMatchSnapshot('error log')
    },
    20 * 1000,
  )

  test(
    'incremental update',
    async () => {
      const { commitIdTable, commitTable } = await buildRepo1({
        repoDir: plainRootDir,
        logger,
        execaOptions: {},
      })

      // Commit E.
      {
        await checkBranch({ ...plainCtx, branchOrCommitId: commitIdTable.E })
        await deleteBranch({ ...plainCtx, branchName: 'main', force: true })
        await createBranch({ ...plainCtx, newBranchName: 'E', branchOrCommitId: commitIdTable.E })
        await checkBranch({ ...plainCtx, branchOrCommitId: 'E' })

        // Test encrypt.
        await gitCipher.encrypt({ catalog, pathResolver })
        expect(await getAllLocalBranches({ ...cryptCtx })).toEqual({
          currentBranch: 'E',
          branches: ['E'],
        })
        expect(await getCommitInTopology({ ...cryptCtx, branchOrCommitId: 'E' })).toEqual([
          { id: cryptCommitIdTable.A, parents: [] },
          { id: cryptCommitIdTable.B, parents: [cryptCommitIdTable.A] },
          { id: cryptCommitIdTable.C, parents: [cryptCommitIdTable.B] },
          { id: cryptCommitIdTable.E, parents: [cryptCommitIdTable.C] },
        ])
        expect(await getCommitWithMessageList({ ...cryptCtx, branchOrCommitIds: ['E'] })).toEqual([
          { id: cryptCommitIdTable.E, message: `#${commitIdTable.E}` },
          { id: cryptCommitIdTable.C, message: `#${commitIdTable.C}` },
          { id: cryptCommitIdTable.B, message: `#${commitIdTable.B}` },
          { id: cryptCommitIdTable.A, message: `#${commitIdTable.A}` },
        ])

        // Test Decrypt
        await gitCipher.decrypt({ pathResolver: bakPathResolver })
        expect(await getAllLocalBranches({ ...bakPlainCtx })).toEqual({
          currentBranch: 'E',
          branches: ['E'],
        })
        expect(await getCommitInTopology({ ...bakPlainCtx, branchOrCommitId: 'E' })).toEqual([
          { id: commitIdTable.A, parents: [] },
          { id: commitIdTable.B, parents: [commitIdTable.A] },
          { id: commitIdTable.C, parents: [commitIdTable.B] },
          { id: commitIdTable.E, parents: [commitIdTable.C] },
        ])
        expect(
          await getCommitWithMessageList({ ...bakPlainCtx, branchOrCommitIds: ['E'] }),
        ).toEqual([
          { id: commitIdTable.E, message: commitTable.E.message },
          { id: commitIdTable.C, message: commitTable.C.message },
          { id: commitIdTable.B, message: commitTable.B.message },
          { id: commitIdTable.A, message: commitTable.A.message },
        ])
      }

      // Commit I
      {
        await createBranch({ ...plainCtx, newBranchName: 'I', branchOrCommitId: commitIdTable.I })

        // Test encrypt.
        await gitCipher.encrypt({ catalog, pathResolver })
        expect(await getAllLocalBranches({ ...cryptCtx })).toEqual({
          currentBranch: 'E',
          branches: ['E', 'I'],
        })
        expect(await getCommitInTopology({ ...cryptCtx, branchOrCommitId: 'I' })).toEqual([
          { id: cryptCommitIdTable.A, parents: [] },
          { id: cryptCommitIdTable.B, parents: [cryptCommitIdTable.A] },
          { id: cryptCommitIdTable.G, parents: [cryptCommitIdTable.B] },
          { id: cryptCommitIdTable.I, parents: [cryptCommitIdTable.G] },
        ])
        expect(await getCommitWithMessageList({ ...cryptCtx, branchOrCommitIds: ['I'] })).toEqual([
          { id: cryptCommitIdTable.I, message: `#${commitIdTable.I}` },
          { id: cryptCommitIdTable.G, message: `#${commitIdTable.G}` },
          { id: cryptCommitIdTable.B, message: `#${commitIdTable.B}` },
          { id: cryptCommitIdTable.A, message: `#${commitIdTable.A}` },
        ])

        // Test Decrypt
        await gitCipher.decrypt({ pathResolver: bakPathResolver })
        expect(await getAllLocalBranches({ ...bakPlainCtx })).toEqual({
          currentBranch: 'E',
          branches: ['E', 'I'],
        })
        expect(await getCommitInTopology({ ...bakPlainCtx, branchOrCommitId: 'I' })).toEqual([
          { id: commitIdTable.A, parents: [] },
          { id: commitIdTable.B, parents: [commitIdTable.A] },
          { id: commitIdTable.G, parents: [commitIdTable.B] },
          { id: commitIdTable.I, parents: [commitIdTable.G] },
        ])
        expect(
          await getCommitWithMessageList({ ...bakPlainCtx, branchOrCommitIds: ['I'] }),
        ).toEqual([
          { id: commitIdTable.I, message: commitTable.I.message },
          { id: commitIdTable.G, message: commitTable.G.message },
          { id: commitIdTable.B, message: commitTable.B.message },
          { id: commitIdTable.A, message: commitTable.A.message },
        ])
      }

      // Commit K
      {
        await createBranch({ ...plainCtx, newBranchName: 'K', branchOrCommitId: commitIdTable.K })
        await checkBranch({ ...plainCtx, branchOrCommitId: 'K' })

        // Test encrypt.
        await gitCipher.encrypt({ catalog, pathResolver })
        expect(await getAllLocalBranches({ ...cryptCtx })).toEqual({
          currentBranch: 'K',
          branches: ['E', 'I', 'K'],
        })
        expect(await getCommitInTopology({ ...cryptCtx, branchOrCommitId: 'K' })).toEqual([
          { id: cryptCommitIdTable.A, parents: [] },
          { id: cryptCommitIdTable.B, parents: [cryptCommitIdTable.A] },
          { id: cryptCommitIdTable.C, parents: [cryptCommitIdTable.B] },
          { id: cryptCommitIdTable.D, parents: [cryptCommitIdTable.C] },
          { id: cryptCommitIdTable.E, parents: [cryptCommitIdTable.C] },
          { id: cryptCommitIdTable.F, parents: [cryptCommitIdTable.E, cryptCommitIdTable.D] },
          { id: cryptCommitIdTable.G, parents: [cryptCommitIdTable.B] },
          { id: cryptCommitIdTable.H, parents: [cryptCommitIdTable.G, cryptCommitIdTable.E] },
          { id: cryptCommitIdTable.I, parents: [cryptCommitIdTable.G] },
          {
            id: cryptCommitIdTable.J,
            parents: [cryptCommitIdTable.I, cryptCommitIdTable.H, cryptCommitIdTable.F],
          },
          { id: cryptCommitIdTable.K, parents: [cryptCommitIdTable.J] },
        ])
        expect(await getCommitWithMessageList({ ...cryptCtx, branchOrCommitIds: ['K'] })).toEqual([
          { id: cryptCommitIdTable.K, message: `#${commitIdTable.K}` },
          { id: cryptCommitIdTable.J, message: `#${commitIdTable.J}` },
          { id: cryptCommitIdTable.I, message: `#${commitIdTable.I}` },
          { id: cryptCommitIdTable.H, message: `#${commitIdTable.H}` },
          { id: cryptCommitIdTable.G, message: `#${commitIdTable.G}` },
          { id: cryptCommitIdTable.F, message: `#${commitIdTable.F}` },
          { id: cryptCommitIdTable.E, message: `#${commitIdTable.E}` },
          { id: cryptCommitIdTable.D, message: `#${commitIdTable.D}` },
          { id: cryptCommitIdTable.C, message: `#${commitIdTable.C}` },
          { id: cryptCommitIdTable.B, message: `#${commitIdTable.B}` },
          { id: cryptCommitIdTable.A, message: `#${commitIdTable.A}` },
        ])

        // Test Decrypt
        await gitCipher.decrypt({ pathResolver: bakPathResolver })
        expect(await getAllLocalBranches({ ...bakPlainCtx })).toEqual({
          currentBranch: 'K',
          branches: ['E', 'I', 'K'],
        })
        expect(await getCommitInTopology({ ...bakPlainCtx, branchOrCommitId: 'K' })).toEqual([
          { id: commitIdTable.A, parents: [] },
          { id: commitIdTable.B, parents: [commitIdTable.A] },
          { id: commitIdTable.C, parents: [commitIdTable.B] },
          { id: commitIdTable.D, parents: [commitIdTable.C] },
          { id: commitIdTable.E, parents: [commitIdTable.C] },
          { id: commitIdTable.F, parents: [commitIdTable.E, commitIdTable.D] },
          { id: commitIdTable.G, parents: [commitIdTable.B] },
          { id: commitIdTable.H, parents: [commitIdTable.G, commitIdTable.E] },
          { id: commitIdTable.I, parents: [commitIdTable.G] },
          {
            id: commitIdTable.J,
            parents: [commitIdTable.I, commitIdTable.H, commitIdTable.F],
          },
          { id: commitIdTable.K, parents: [commitIdTable.J] },
        ])
        expect(
          await getCommitWithMessageList({ ...bakPlainCtx, branchOrCommitIds: ['K'] }),
        ).toEqual([
          { id: commitIdTable.K, message: commitTable.K.message },
          { id: commitIdTable.J, message: commitTable.J.message },
          { id: commitIdTable.I, message: commitTable.I.message },
          { id: commitIdTable.H, message: commitTable.H.message },
          { id: commitIdTable.G, message: commitTable.G.message },
          { id: commitIdTable.F, message: commitTable.F.message },
          { id: commitIdTable.E, message: commitTable.E.message },
          { id: commitIdTable.D, message: commitTable.D.message },
          { id: commitIdTable.C, message: commitTable.C.message },
          { id: commitIdTable.B, message: commitTable.B.message },
          { id: commitIdTable.A, message: commitTable.A.message },
        ])
      }

      expect(logMock.getIndiscriminateAll()).toMatchSnapshot('error log')
    },
    20 * 1000,
  )
})
