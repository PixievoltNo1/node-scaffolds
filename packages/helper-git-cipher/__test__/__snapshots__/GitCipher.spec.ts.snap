// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitCipher from scratch: error log 1`] = `
[
  [
    "[safeExeca] failed to run:",
    "git",
    "merge 06d25f06c6cd40756bf61624f1ee37bf014ec6d0 28fc4e74bf3bc436c21774dfc9947d60116d9716 -m 'H -- Merge E and G <conflict> (b2,c3)'",
  ],
]
`;

exports[`GitCipher incremental update: error log 1`] = `
[
  [
    "[safeExeca] failed to run:",
    "git",
    "merge 06d25f06c6cd40756bf61624f1ee37bf014ec6d0 28fc4e74bf3bc436c21774dfc9947d60116d9716 -m 'H -- Merge E and G <conflict> (b2,c3)'",
  ],
]
`;
